addon/-private/system/normalize-link.ts

```diff
@@ -10,7 +18,7 @@
   @return {Object|null}
   @for DS
 */
-export default function _normalizeLink(link) {
+export default function _normalizeLink(link: Link | string): NormalizedLink {
   switch (typeof link) {
     case 'object':
       return link;
```

Comment:
an exhaustive switch should be used here http://ideasintosoftware.com/exhaustive-switch-in-typescript/

`NormalizedLink` has the potential to be null -- what value for `Link|string` could be passed as an argument, which would result in `null` being returned?

Reply:
done, and yes, null => null, but only because `typeof null === 'object'`

Reply:
I'm less concerned about `typeof` than I am about type equivalence in a `"strictNullChecks": true` situation. 

```ts
const myObj1: object = null; // ❌
const myObj2: {} = null; // ❌
```

Reply:
I'm confused by your link about exhaustive switch, as it uses an enum, and has nothing to do with types.

I can explicitly check for null, which would be more intuitive than what is going on now

Reply:
The link is getting at the fact that you can make TS give you a *type* error if you enumerate the cases you expect, so that if you expand that number of cases later and fail to cover it, you'll *know* about it. The easiest (though not only) way to do that is with a string-valued enum; you could also do it simply with a type alias enumerating the cases.

Reply:
I don't know how to apply that here.
sorry for not understanding :(

Reply:
No worries! It's complicated and not *super* well-explained in the docs if you don't already have a frame for what's going on. I will try to write up a good explanation as well as a specific recommendation for how to implement it in this PR tomorrow!